inference-engine/src/cldnn_engine/cldnn_engine.cpp

```diff
@@ -703,11 +702,71 @@ Parameter clDNNEngine::GetMetric(const std::string& name, const std::map<std::st
     } else if (name == METRIC_KEY(RANGE_FOR_STREAMS)) {
         std::tuple<unsigned int, unsigned int> range = std::make_tuple(1, 2);
         IE_SET_METRIC_RETURN(RANGE_FOR_STREAMS, range);
+    } else if (name == GPU_METRIC_KEY(MAX_BATCH_SIZE)) {
+        const auto& config = _impl->m_configs.GetConfig(device_id);
+        auto n_streams = config.throughput_streams;
+        auto available_device_mem = device_info.max_global_mem_size;
+        int64_t max_batch_size = 0;
+
+        if (options.find("CNN_NETWORK") == options.end()) {
+            throw std::runtime_error("No CNN_NETWORK option given!");
+        }
+        if (options.find("GPU_THROUGHPUT_STREAMS") != options.end()) {
+            n_streams = options.find("GPU_THROUGHPUT_STREAMS")->second.as<uint16_t>();
+        }
+        if (options.find("AVAILABLE_DEVICE_MEM_SIZE") != options.end()) {
+            available_device_mem = options.find("AVAILABLE_DEVICE_MEM_SIZE")->second.as<int64_t>();
+            if (available_device_mem < 0)
+                throw std::runtime_error("No available device mem");
+        }
+
+        auto network = options.find("CNN_NETWORK")->second.as<InferenceEngine::CNNNetwork*>();
+
+        auto input_shapes = network->getInputShapes();
+        std::string input_name;
+        SizeVector input_shape;
+        std::tie(input_name, input_shape) = *input_shapes.begin();
+        size_t base_batch_size = 16; // empirically decided for DG1
+
+        auto engine = clDNNEngineFactory::create(config, iter->second, nullptr, true);
+        std::shared_ptr<Program> program;
+
+        if (options.find("BASE_BATCH_SIZE") != options.end() && base_batch_size != options.find("BASE_BATCH_SIZE")->second.as<int32_t>()) {
+            base_batch_size = options.find("BASE_BATCH_SIZE")->second.as<int32_t>();
+        }
+
+        if (base_batch_size != input_shape[0]) {
+            auto cloned_network = InferenceEngine::details::cloneNetwork(*network);
+            auto input_shapes = cloned_network.getInputShapes();
+            std::string input_name;
+            SizeVector input_shape;
+            std::tie(input_name, input_shape) = *input_shapes.begin();
+            input_shape[0] = base_batch_size;
```

Comment:
this NOT the safe way to reshape network by the batch!
first of all it assumes single input only (rather then do a loop), secondly it should check the layout is really a N*** (like NCHW), before patching zero-th dim.
While the auto-batching branch is doing that: https://github.com/myshevts/openvino/blob/auto-batch-master/inference-engine/src/auto_batch/auto_batch.cpp#L35 it is not fully ok either....
**@jane-intel, we all are pending the robust reshape-by-batch functionality (CVS-51744) to avoid home-baked reshaping codes like these**

Finally the

Reply:
Okay, so the work done for 51744 is expected to be used here. Seems that it will start from this week.. then we have the dependency.

Reply:
I guess you can create a leftover ticket for youself in the Epic (and add a TODO comment to the code) on switching to the "right" reshaping-by-batch and continue with this PR without blocking

Reply:
In cldnn the assumed layout is NCHW or NCDHW, and if the dim is less than 4 it is filling 1. 
https://github.com/openvinotoolkit/openvino/blob/c8b44b275cb11eed74e6ded64ab5f2afbf1cd21d/inference-engine/src/cldnn_engine/cldnn_common_utils.h#L16
This assumption is basically being used in cldnn plugin, so could I apply this assumption here too? 
i.e., Here
check dims and if dim is < 4 => fill 1 and reshape.

Reply:
@vladimir-paramuzov

Reply:
for ngraph::Function reshape this cldnn assumption won't work, as it will change rank of tensors which may lead to mismatch in shape propagation logic.

Basically, you just need to change the logic above to iterate over all inputs and find position of batch dimension using the layout, and then modify batch dimension

Reply:
There are 5 layouts C, NC, CHW, NCHW, NCDHW where the layouts starting with N the batch is 0th index and there is no batch for the others. If we cannot change the rank (e.g., adding additional batch dim for "CHW" layout), is it okay to 
1) set base batch 1 for C/ CHW 
2) for other layouts, reshape 0th dim to 16?

Reply:
Kind of. If layout doesn't have batch dimension at all, then you just keep the shape as is. If batch dimension is present in layout, then you need to check the position of batch and change it in the shape. This will allow to handle any layout (NHWC, NCHW, CHW, etc)

Reply:
Then for such cases, autobatch will not be applied too, right?

Reply:
Fixed the reshape

Reply:
the auto-batching will use the transformation that will be doing safe reshape by the batch, it will skip non-reshapabe-by-batch networks (and use the original batch1 network)

Reply:
This getmaxbatch is also reshaping networks with batch dimension only. If the layout is not uisng batch dim, it will use batch size 1.